text/0000-read-buf.md

```diff
@@ -0,0 +1,624 @@
+        // Get access to the unwritten part of the buffer, making sure it has been fully initialized. Since `ReadBuf`
+        // tracks the initialization state of the buffer, this is "free" after the first time it's called.
+        let unwritten: &mut [u8] = buf.initialize_unwritten();
+
+        // Fill the whole buffer with some nonsense.
+        for (i, byte) in unwritten.iter_mut().enumerate() {
+            *byte = i as u8;
+        }
+
+        // And indicate that we've written the whole thing.
+        let len = unwritten.len();
+        buf.add_written(len);
+
+        Ok(())
+    }
+}
+```
+
+An unsafe `Read` implementation:
+
+```rust
+impl Read for TcpStream {
+    fn read_buf(&mut self, buf: &mut ReadBuf<'_>) -> io::Result<()> {
+        unsafe {
+            // Get access to the unwritten part of the buffer, without initializing it. This method is unsafe; we are
+            // responsible for ensuing that we don't "de-initialize" portions of it that have previously been
+            // initialized.
+            let unwritten: &mut [MaybeUninit<u8>] = buf.unwritten_mut();
+
+            // We're just delegating to the libc read function, which returns an `isize`. The return value indicates
+            // an error if negative and the number of bytes read otherwise.
+            let nread = libc::read(self.fd, unwritten.as_mut_ptr().cast::<libc::c_void>(), unwritten.len());
+
+            if nread < 0 {
+                return Err(io::Error::last_os_error());
+            }
+
+            let nread = nread as usize;
+            // If the read succeeded, tell the buffer that the read-to portion has been initialized. This method is
+            // unsafe; we are responsible for ensuring that this portion of the buffer has actually been initialized.
+            buf.assume_initialized(nread);
+            // And indicate that we've written the bytes as well. Unlike `assume_initialized`, this method is safe,
+            // and asserts that the written portion of the buffer does not advance beyond the initialized portion of
+            // the buffer. If we didn't call `assume_initialized` above, this call could panic.
+            buf.add_written(nread);
+
+            Ok(())
+        }
+    }
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+```rust
+/// A wrapper around a byte buffer that is incrementally written to and initialized.
+///
+/// This type is a sort of "double cursor". It tracks three regions in the buffer: a region at the beginning of the
+/// buffer that has been logically written to, a region that has been initialized at some point but not yet logically
+/// written to, and a region at the end that is fully uninitialized. The written-to region is guaranteed to be a
+/// subset of the initialized region.
+pub struct ReadBuf<'a> {
+    buf: &'a mut [MaybeUninit<u8>],
+    written: usize,
+    initialized: usize,
+}
```

Comment:
At the risk of opening a can of worms, this feels *very similar* to [arrayvec](https://crates.io/crates/arrayvec). This makes me wonder a few points:

1. Could this be made generic over `u8`, so it could be used for other similar "partial initialization" cases?

1. Could there be an owning variant of this so that we could do something like

    ```rust
    let mut buf = ReadBufOwned::<const 8192>();
    ```

    I'm using unstable const generics here, sadly, but I think the ergonomics are nice. 

Because of #2, I'd also like to bikeshed about the name <code>Read<b>Buf</b></code> a little. We have `Path` and `PathBuf`, where `*Buf` is the owned version, so I originally wanted to type `ReadBufBuf`...

Reply:
> 1\. Could this be made generic over u8, so it could be used for other similar "partial initialization" cases?

It *could* be made generic, but I'm not sure the equivalence is really tight enough. The API of ReadBuf is centered entirely around progressive initialization, which doesn't necessarily align with the API for a general purpose vec-style data structure.

> 2\. Could there be an owning variant of this so that we could do something like

That's covered briefly in the future possibilities section.

Reply:
> That's covered briefly in the future possibilities section.

Are you referring to this section?

>> There is probably some kind of abstraction that could be defined to encapsulate that logic.

If so, perhaps that section could be extended a bit; I've re-read it a few times and still can't see a connection to an owning version.

Reply:
I, too, saw some avenues for possible generalizations: currently one is unable to use `Vec` in `no_std` crates. It's preferable for a library to let the caller to decide the "backing store" of a buffer. However, in cases where there is input-dependent or otherwise unknown amount of data stored in the buffer, it's generally desirable to have a type that keeps track the "written part" and "uninitialized part", for both ergonomics (less book-keeping for the user) and performance (skipping initialization). In this sense, `Vec` is pretty awesome, but it comes at the cost of an allocation.

It's also desirable to have such type as a part of the "common vocabulary" of the ecosystem, but alas, std doesn't have a type that fits the bill at the moment.

Thus, I think it would be valuable to provide a type that is generalized over the contained type, and keeps track the "written"/"uninitialized" split. We could then only allow `Read`ing, but `Write`ing to it, and consider it as a generic buffer type for some other uses, such as a replacement for `Vec` for libraries that need to work in `no_std` contexts.

Reply:
A Vec-like abstraction that uses a fixed-size slice of uninitialized memory as backing storage would indeed be a useful abstraction, but requires basically reimplementing the entire Vec if done outside the standard library. I've written at length about this issue [here](https://internals.rust-lang.org/t/pre-rfc-fixed-capacity-view-of-vec/8413). That writeup may be a bit outdated, but I'm glad to elaborate if you believe this is relevant to this discussion.